### PR TITLE
[SVS-9] Implemented ErrorListInfoBarController

### PR DESCRIPTION
**Pull request by eyalantzMS**
Implemented the controller that will show the info bar and will handle the Update command.

I've considered just using the binding and connection workflows instead of driving the experience via the section, but decided not to do it since it could introduce various race conditions and avoiding them would require increasing the overall complexity which i preferred to avoid.
The approach used drives the flow using the WPF commands by passing the right arguments to those commands, so essentially exactly the same as the user would have done, so no added risk.

Misc:
1. BindingController - removed dead code that was checked in part of SLVS-87 + some other minor change to be consisted with the ConnectionController
2. ConnectionController/Workflow - Simplified the code (was getting a delegate that could have been directly called from within the workflow)
3. Added another telemetry to the fix-conflicts: one that records the number of times we showed the UX in addition to the existing fix-conflicts command telemetry. Now we should be able to see if the users trust that feature :)

### CODE REVIEW

File: src/Integration.UnitTests/Connection/ConnectionWorkflowTests.cs [hunk 1]
```diff
@@ -66,7 +66,7 @@ public void ConnectionWorkflow_ConnectionStep_SuccessfulConnection()
             var projects = new ProjectInformation[] { new ProjectInformation { Key = "project1" } };
             this.sonarQubeService.ReturnProjectInformation = projects;
             bool projectChangedCallbackCalled = false;
-            ConnectedProjectsCallback projectsChanged = (c, p) =>
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/Connection/ConnectionWorkflowTests.cs]: Removed this delegate sonce SetProject could be called directly (and tested directly)

File: src/Integration.UnitTests/LocalServices/ErrorListInfoBarControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+using SonarLint.VisualStudio.Integration.Resources;
+using SonarLint.VisualStudio.Integration.Service;
+using SonarLint.VisualStudio.Integration.TeamExplorer;
+using SonarLint.VisualStudio.Integration.WPF;
+using System;
+using System.ComponentModel.Composition.Primitives;
+using System.Linq;
+using System.Windows.Threading;
+
+namespace SonarLint.VisualStudio.Integration.UnitTests
```
Comment (eyalantzMS) [inline src/Integration.UnitTests/LocalServices/ErrorListInfoBarControllerTests.cs]: There's twice as much of code here as in the test subject just due to all the edge cases, partial list:
1) clicking Update twice
2) The user disconnected after the info bar have been shown
3) The user connected to a different server after the info bar have been shown
4) The user bound to a different project after the info bar have been shown
5) Build/Debug started (the underlying commands will return false when asked if can execute)

File: src/Integration.Vsix/Telemetry/GeneratedCode/SqmCommandFacade.cs [hunk 3]
```diff
@@ -157,12 +157,30 @@ public static void FixConflictsCommand()
         }
```
Comment (eyalantzMS) [inline src/Integration.Vsix/Telemetry/GeneratedCode/SqmCommandFacade.cs]: Generated code

File: src/Integration.Vsix/Telemetry/GeneratedCode/SqmOleCommandTarget.cs [hunk 4]
```diff
@@ -33,7 +33,9 @@ internal enum SonarLintSqmCommandIds
         ToggleShowAllProjectsCommandCommandId = 0x306,
         DontWarnAgainCommandCommandId = 0x307,
         FixConflictsCommandCommandId = 0x308,
-        UpdateBindingCommandFromErrorListCommandId = 0x400,
+        FixConflictsShowCommandId = 0x309,
```
Comment (eyalantzMS) [inline src/Integration.Vsix/Telemetry/GeneratedCode/SqmOleCommandTarget.cs]: Generated code

File: src/Integration.Vsix/Telemetry/GeneratedCode/SqmOleCommandTarget.cs [hunk 5]
```diff
@@ -96,7 +98,9 @@ public static bool IsSqmCommand(Guid commandGroup, int commandId)
                     case SonarLintSqmCommandIds.ToggleShowAllProjectsCommandCommandId:
                     case SonarLintSqmCommandIds.DontWarnAgainCommandCommandId:
                     case SonarLintSqmCommandIds.FixConflictsCommandCommandId:
-                    case SonarLintSqmCommandIds.UpdateBindingCommandFromErrorListCommandId:
+                    case SonarLintSqmCommandIds.FixConflictsShowCommandId:
```
Comment (eyalantzMS) [inline src/Integration.Vsix/Telemetry/GeneratedCode/SqmOleCommandTarget.cs]: Generated code

File: src/Integration.Vsix/Telemetry/GeneratedCode/SqmVsCommands.vsct [hunk 6]
```diff
@@ -95,11 +95,27 @@
                     </Strings>
                 </Button>
   
-                <Button guid="guidCmdSet" id="UpdateBindingCommandFromErrorList" priority="0x00" type="Button">
+                <Button guid="guidCmdSet" id="FixConflictsShow" priority="0x00" type="Button">
```
Comment (eyalantzMS) [inline src/Integration.Vsix/Telemetry/GeneratedCode/SqmVsCommands.vsct]: Generated code

File: src/Integration.Vsix/Telemetry/SonarLintSqmData.ttinclude [hunk 7]
```diff
@@ -39,8 +39,10 @@ IEnumerable<Tuple<string, string, string>> commands = new ReadOnlyCollection <Tu
         Tuple.Create("ToggleShowAllProjectsCommand",		"0x306", "SonarLint.VisualStudio ToggleShowAllProjectsCommand was called"),
         Tuple.Create("DontWarnAgainCommand",				"0x307", "SonarLint.VisualStudio DontWarnAgainCommand was called"),
         Tuple.Create("FixConflictsCommand",					"0x308", "SonarLint.VisualStudio FixConflictsCommand was called"),
+        Tuple.Create("FixConflictsShow",					"0x309", "SonarLint.VisualStudio FixConflicts UX popped up"),
         // Info bar
-        Tuple.Create("UpdateBindingCommandFromErrorList",	"0x400", "SonarLint.VisualStudio ErrorList info bar Update button was clicked"),
+        Tuple.Create("ErrorListInfoBarShow",				"0x400", "SonarLint.VisualStudio ErrorList info bar UX popped up"),
```
Comment (eyalantzMS) [inline src/Integration.Vsix/Telemetry/SonarLintSqmData.ttinclude]: Added telemetry, decided to change the name and the id if the Update command telemetry. Should not matter since nothing was firing it up until this PR.

File: src/Integration/Persistence/BoundSonarQubeProjectExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+//-----------------------------------------------------------------------
+
+using SonarLint.VisualStudio.Integration.Service;
+using System;
+
+namespace SonarLint.VisualStudio.Integration.Persistence
+{
+    internal static class BoundSonarQubeProjectExtensions
+    {
+        public static ConnectionInformation CreateConnectionInformation(this BoundSonarQubeProject binding)
```
Comment (eyalantzMS) [inline src/Integration/Persistence/BoundSonarQubeProjectExtensions.cs]: Pre-existing code that was extracted to an extension method.

File: src/Integration/TeamExplorer/ProjectViewModel.cs [hunk 9]
```diff
@@ -18,6 +18,9 @@ internal class ProjectViewModel : ViewModelBase
         private readonly ContextualCommandsCollection commands = new ContextualCommandsCollection();
         private bool isBound;
 
+        // Ordinal comparer should be good enough: http://docs.sonarqube.org/display/SONAR/Project+Administration#ProjectAdministration-AddingaProject
+        public static readonly StringComparer KeyComparer = StringComparer.Ordinal;
```
Comment (eyalantzMS) [inline src/Integration/TeamExplorer/ProjectViewModel.cs]: Existed before, made public static for reuse in other places

File: src/Integration/Resources/Strings.resx [hunk 10]
```diff
@@ -547,4 +547,42 @@ This is the general format to use when writing errors to output window or when t
   <data name="SolutionIsClosed" xml:space="preserve">
     <value>The operation requires an open solution.</value>
   </data>
+  <data name="SonarLintInfoBarUpdateCommandIsBusyRetry" xml:space="preserve">
+    <value>SonarLint: Could not update the binding since the command is not available at the moment. Please try again and make sure that not building or debugging.</value>
+    <comment>Output window message</comment>
+  </data>
+  <data name="SonarLintInfoBarUnboundProjectsMessage" xml:space="preserve">
+    <value>SonarLint: One or more RuleSet is out of date or not linked to the SonarQube profile-based RuleSet (see General Output for affected projects)</value>
```
Comment (mjcheetham) [inline src/Integration/Resources/Strings.resx]: > SonarLint: One or more rule sets are out of date or not linked to the SonarQube quality profile rule set (see General Output for affected projects)

File: src/Integration/Resources/Strings.resx [hunk 11]
```diff
@@ -547,4 +547,42 @@ This is the general format to use when writing errors to output window or when t
   <data name="SolutionIsClosed" xml:space="preserve">
     <value>The operation requires an open solution.</value>
   </data>
+  <data name="SonarLintInfoBarUpdateCommandIsBusyRetry" xml:space="preserve">
+    <value>SonarLint: Could not update the binding since the command is not available at the moment. Please try again and make sure that not building or debugging.</value>
```
Comment (mjcheetham) [inline src/Integration/Resources/Strings.resx]: > SonarLint: Could not update the binding because the command is not currently available. Please try again and ensure that the solution is not building or currently being debugged.

File: src/Integration/Resources/Strings.resx [hunk 12]
```diff
@@ -547,4 +547,42 @@ This is the general format to use when writing errors to output window or when t
...
+    <comment>Output window message
+{0} the number of unbound projects
+{1} projects unique names</comment>
+  </data>
+  <data name="SonarLintNoUnboundProjectWereFound" xml:space="preserve">
+    <value>SonarLint: No unbound projects were found.</value>
+    <comment>Output window message</comment>
+  </data>
+  <data name="SonarLintInfoBarUpdateCommandRetryNoActiveSection" xml:space="preserve">
+    <value>The SonarQube active section was closed and we cannot proceed with the update. Please try again.</value>
```
Comment (mjcheetham) [inline src/Integration/Resources/Strings.resx]: "active section" is pretty technical.
How about:
> "The SonarQube Team Explorer page was closed..."

File: src/TestInfrastructure/Framework/ConfigurableStateManager.cs [hunk 13]
```diff
@@ -126,6 +126,7 @@ private void VerifyActiveSection()
 
         public void InvokeBusyChanged(bool value)
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableStateManager.cs]: Now it's doing two things, `SetAndInvokeBusyChanged`?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 14]
```diff
@@ -21,22 +39,427 @@ public ErrorListInfoBarController(IHost host)
...
+                return this.bindingInformationProvider.IsSolutionBound();
+            }
+        }
+
+        private void InvokeWhenIdle(Action action)
+        {
+            Debug.Assert(action != null);
+
+            this.host.UIDispatcher.BeginInvoke(
+                  DispatcherPriority.ContextIdle,
```
Comment (mjcheetham) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Realistically when will we be 'idle'? What is the definition of 'idle'?
If I open a solution and I start working, running tests, debugging, adding classes, using analyser fixes, typing, is that idle? My worry is we won't run until they go for tea.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 15]
```diff
@@ -21,22 +39,427 @@ public ErrorListInfoBarController(IHost host)
...
+            // when the delegate will execute, and we should handle those cases
+            if (this.IsActiveSolutionBound)
+            {
+                this.InvokeWhenIdle(this.ProcessSolutionBinding);
+            }
         }
         #endregion
 
         #region Non-public API
+        private void AssertOnUIThread()
```
Comment (mjcheetham) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: `[Conditional("DEBUG")]`

File: src/Integration/State/StateManager.cs [hunk 16]
```diff
@@ -189,8 +189,7 @@ private void RestoreBoundProject(ServerViewModel serverViewModel)
                 return;
             }
 
-            // Ordinal comparer should be good enough: http://docs.sonarqube.org/display/SONAR/Project+Administration#ProjectAdministration-AddingaProject
-            ProjectInformation boundProject = serverViewModel.Projects.FirstOrDefault(pvm => StringComparer.Ordinal.Equals(pvm.Key, this.BoundProjectKey))?.ProjectInformation;
+            ProjectInformation boundProject = serverViewModel.Projects.FirstOrDefault(pvm => ProjectViewModel.KeyComparer.Equals(pvm.Key, this.BoundProjectKey))?.ProjectInformation;
```
Comment (mjcheetham) [inline src/Integration/State/StateManager.cs]: Why is the `KeyComparer` on `ProjectViewModel`? Shouldn't it really be on `ProjectInformation`?

File: src/Integration/TeamExplorer/ProjectViewModel.cs (hunk 9 above)
Comment (mjcheetham) [inline src/Integration/TeamExplorer/ProjectViewModel.cs]: Why is the `KeyComparer` on `ProjectViewModel`? Shouldn't it really be on `ProjectInformation`?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 14 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: ContextIdle - Operations are processed after background operations have completed.

The context here is the UI thread, so that slot with be sooner then you think. If according to user feedback this will not be soon enough in some case we can do fancier things, until then this seems simple enough to work in practice.

File: src/Integration/TeamExplorer/ProjectViewModel.cs (hunk 9 above)
Comment (eyalantzMS) [inline src/Integration/TeamExplorer/ProjectViewModel.cs]: Yep.

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 17]
```diff
@@ -5,13 +5,31 @@
...
+using SonarLint.VisualStudio.Integration.Service;
+using SonarLint.VisualStudio.Integration.TeamExplorer;
 using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Windows.Threading;
 
 namespace SonarLint.VisualStudio.Integration
 {
     internal class ErrorListInfoBarController : IErrorListInfoBarController
```
Comment (mjcheetham) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Should this not be `IDisposable` as well as `ILocalService`?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 18]
```diff
@@ -21,22 +39,428 @@ public ErrorListInfoBarController(IHost host)
...
+        {
+            this.ClearCurrentInfoBar();
+        }
+
+        private void CurrentErrorWindowInfoBar_ButtonClick(object sender, EventArgs e)
+        {
+            if (this.currentErrorWindowInfoBarHandlingClick)
+            {
+                // Info bar doesn't expose a way to disable the command
+                // and since the code in asynchronous the user can click
```
Comment (mjcheetham) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: is*

File: src/Integration/LocalServices/ErrorListInfoBarController.cs [hunk 19]
```diff
@@ -21,22 +39,428 @@ public ErrorListInfoBarController(IHost host)
...
+                else
+                {
+                    // Start from connection state
+                    this.RefreshConnection();
+                }
+            }
 
+            private void RefreshConnection()
+            {
+                this.WhenNotBusy(this.ExecuteRefreshCommand);
```
Comment (mjcheetham) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: Is there a need for all these calls to `WhenNotBusy`? We can end up nesting a lot:
`ConnectAndBind -> WNB -> Start -> Refresh -> WNB -> ExecRefreshCmd -> UpdateBinding -> WNB -> OnFinished`

File: src/Integration/Persistence/BoundSonarQubeProjectExtensions.cs (hunk 8 above)
Comment (mjcheetham) [inline src/Integration/Persistence/BoundSonarQubeProjectExtensions.cs]: Tests?

File: src/TestInfrastructure/Framework/ConfigurableInfoBar.cs [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+        public string Message { get; }
+        public string ButtonText { get; }
+        public ImageMoniker Image { get; }
+
+        public void AssertClosedCalled(int expectedNumberOfTimes)
+        {
+            Assert.AreEqual(expectedNumberOfTimes, this.closedCalled, $"{nameof(Close)} was called unexpected number of times");
+        }
+
+        public void SimulatButtonClickEvent()
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableInfoBar.cs]: Simulate*

File: src/TestInfrastructure/Framework/ConfigurableInfoBar.cs [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            this.ButtonClick?.Invoke(this, EventArgs.Empty);
+        }
+
+        public void SimulatClosedEvent()
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableInfoBar.cs]: Simulate*

File: src/TestInfrastructure/Framework/ConfigurableInfoBar.cs [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            this.Closed?.Invoke(this, EventArgs.Empty);
+        }
+
+        public void VerifyAllEventsUnregistered()
+        {
+            Assert.IsNull(this.ButtonClick, $"{nameof(this.ButtonClick)} event was remained registered");
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableInfoBar.cs]: > ..event <s>was</s> remained registered

File: src/TestInfrastructure/Framework/ConfigurableSolutionBindingInformationProvider.cs [hunk 23]
```diff
@@ -1,4 +1,4 @@
-//-----------------------------------------------------------------------
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableSolutionBindingInformationProvider.cs]: Added a BOM?

File: src/Integration/LocalServices/ErrorListInfoBarController.cs (hunk 19 above)
Comment (eyalantzMS) [inline src/Integration/LocalServices/ErrorListInfoBarController.cs]: The longest nested chain is when already connected, in other cases since it will be busy it will break the nesting and start a new chain once no longer busy.

File: src/Integration/Persistence/BoundSonarQubeProjectExtensions.cs (hunk 8 above)
Comment (eyalantzMS) [inline src/Integration/Persistence/BoundSonarQubeProjectExtensions.cs]: Was covered implicitly previously, but its worth adding explicit tests.

File: src/TestInfrastructure/Framework/ConfigurableSolutionBindingInformationProvider.cs [hunk 24]
```diff
@@ -1,4 +1,5 @@
-﻿//-----------------------------------------------------------------------
+﻿
```
Comment (mjcheetham) [inline src/TestInfrastructure/Framework/ConfigurableSolutionBindingInformationProvider.cs]: There should be no changes to this file.
Might be better to just revert the file, or remove the extra starting new-line and use VSCode to save the file as UTF-8 (_w/o_ BOM).
image

**Comment by mjcheetham** [comment]
Other than reverting the encoding change to `ConfigurableSolutionBindingInformationProvider.cs`, LGTM.
